[DON'T MERGE] Work on the feature/cuda branch so far

eyalroz opened the pull request:
This PR is not intended for merging, but for inspecting/commenting on the work so far.

What I/we have achieved is:

* The library - with some input formats and some destinations removed - can now be compiled by NVCC when included from a `.cu` file.
* Dropped all (?) dependencies on non-device-side-usable parts of the C++ standard library.
* Made sure all (?) functions and methods can run on both host-side and device-side code. Usually this is merely by decorating them with `__host__ __device__` (or `__hd__` for short); sometimes actual changes were necessary.
* Super-ugly handling of the inability of GPU device-side code to throw exceptions.

robhz786 commented on include/strf/detail/cesu8.hpp:
```diff
@@ -9,58 +9,60 @@
 #include <algorithm>
 #include <cstring>
 
```
Actually this header is not being used. It could be removed.

robhz786 commented on include/strf/detail/printers_tuple.hpp:
```diff
@@ -45,7 +47,7 @@ class simple_tuple_impl<std::index_sequence<I...>, T...>
     template <typename ... Args>
     constexpr explicit simple_tuple_impl(simple_tuple_from_args, Args&& ... args)
-        : detail::indexed_obj<I, T>{args}...
+        : indexed_obj<I, T>(args)...
```
It is necessary to also change `indexed_obj`, otherwise GCC fail to compile. 

```
template <std::size_t I, typename T>
struct indexed_obj
{
    explicit constexpr indexed_obj(const T& _obj)
        : obj(_obj)
    {
    }
    constexpr indexed_obj(const indexed_obj&) = default;
    constexpr indexed_obj(indexed_obj&&) = default;
    
    T obj;
};
```

eyalroz commented on include/strf/detail/char_traits.hpp:
```diff
@@ -0,0 +1,575 @@
+#define STRF_CONSTEXPR_SINCE_CXX17 __hd__ constexpr
+#else
+#define STRF_CONSTEXPR_SINCE_CXX17 __hd__ constexpr
+#endif
```
But I think some of the char traits are only constexpr since C++17. Still, I can drop this if you like.

robhz786 replied:
`STRF_CONSTEXPR_CHAR_TRAITS` is only defined in C++17. And some versions MSVC don't define `__cplusplus` correclty.

robhz786 replied:
Fell free to add `__hd__` in the value of `STRF_CONSTEXPR_CHAR_TRAITS`

eyalroz commented on include/strf/detail/char_traits.hpp:
```diff
@@ -0,0 +1,575 @@
+    for (; n; --n, ++s)
+        assign(*s, a);
+    return r;
+}
```
@robhz786 :  I copied this from  libc++'s standard char traits. If there's any optimization - it's done by the compiler, not by library code directly.

Having said that - I also agree that it makes sense to factor this out. I'll do the factoring-out on my feature branch and drop the char_traits.

eyalroz replied:
Also, `memcpy()` in kernels works, but it's a [slow char-at-a-time loop](https://stackoverflow.com/q/10456728/1593077).

eyalroz commented on include/strf/detail/define_specifiers.hpp:
```diff
@@ -0,0 +1,58 @@
+#ifndef __fd__
+#define __fd__  __forceinline__ __device__
+#endif
+
```
Ok.

eyalroz replied:
Done (on my repo).

eyalroz commented:
Done addressing all comments here. Closing this PR; will open another one after I've added more testing code (and perhaps uncommented some input sources.
